site/content/blog/2022-03-22-schema-migrations.md

```diff
@@ -0,0 +1,126 @@
+    Close() error
+    Apply(migration *models.Migration, saveVersion bool) error
+    AppliedMigrations() ([]*models.Migration, error)
+    SetConfig(key string, value interface{}) error
+}
+```
+
+Morph engine expects the migration scripts to be parsed and accessible from an interface: `source.Source` which is a minimal interface as:
+
+```Go
+type Source interface {
+    Migrations() (migrations []*models.Migration)
+}
+```
+
+The `models.Migration` is a simple struct that contains migration name, version, direction, and the migration statements as an `io.ReadCloser` interface can be a file essentially. The `models.Migration` struct can be initialized from the files and the built-in file source expects files to have a specific format.
+
+The migrations files should have up and down versions. The tool expects each migration to be reversible, and the naming of the migration should be in the following form:
+
+```none
+{order_number}_{migration_name}.{(up|down).sql}
+```
+
+The first part will be used to determine the order in which the migrations should be applied and the next part until the `up|down.sql` suffix will be the migration name.
+
+For example:
+
+```none
+0000000001_create_user.up.sql
+0000000001_create_user.down.sql
+```
+
+Apart from the essential arguments, morph can be initialized with a locking option which enables a lock mechanism (to mimic advisory locking) that’s been builtin with the tool. It is being used for cases where multiple applications could execute the migrations at the same time. If the database driver implements the following interface, morph can be used safely in multi-application deployments:
+
+```Go
+type Locker interface {
+    Lock(context.Context) error
+    Unlock() error
+}
+```
+
+It essentially allows an instance to create a mutex value with expiration in the database and refreshes it until the instance finishes applying migrations. Other instances will keep waiting until the mutex is either removed from the database or get expired. The main difference here is that locking is not used for a single statement but it’s being used for the entire migration process. And this mechanism creates another level above sessions hence it wouldn’t fail if the advisory locks are bound to the sessions.
+
+The morph CLI provides the same functionality from the command line. Feel free to install and experiment as we think the command discovery should be sufficient to start using it!
+
+## Summary
+
+Schema migrations enable versioning as the schema evolves. Making schema changes to databases is serious business as it means changing your customers' data structures. And that always comes with some risks. We consider it is a good practice to have easy recovery paths and to handle changes as a unit. We developed [morph](https://github.com/mattermost/morph/) to achieve these goals. Even though it’s not mature enough to be used in many different application types, we are continuously improving it as requirements change.
```

Comment:
At this point I'm completely sold on morph and I want to use it, but then you say... 

> Even though it’s not mature enough to be used in many different application types

And I'm like, whut? So what application types should I use and not use this for?

Reply:
Softened the statement here :)

Reply:
Yep, better now :)